core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/LeanBalancer.scala

```diff
@@ -0,0 +1,104 @@
+import org.apache.openwhisk.spi.SpiLoader
+import org.apache.openwhisk.utils.ExecutionContextFactory
+import pureconfig._
+import org.apache.openwhisk.core.entity.size._
+
+import scala.concurrent.Future
+
+/**
+ * Lean loadbalancer implemetation.
+ *
+ * Communicates with Invoker directly without Kafka in the middle. Invoker does not exist as a separate entity, it is built together with Controller
+ * Uses LeanMessagingProvider to use in-memory queue instead of Kafka
+ */
+class LeanBalancer(config: WhiskConfig,
+                   feedFactory: FeedFactory,
+                   controllerInstance: ControllerInstanceId,
+                   implicit val messagingProvider: MessagingProvider = SpiLoader.get[MessagingProvider])(
+  implicit actorSystem: ActorSystem,
+  logging: Logging,
+  materializer: ActorMaterializer)
+    extends CommonLoadBalancer(config, feedFactory, controllerInstance) {
+
+  /** Loadbalancer interface methods */
+  override def invokerHealth(): Future[IndexedSeq[InvokerHealth]] = Future.successful(IndexedSeq.empty[InvokerHealth])
+  override def clusterSize: Int = 1
+
+  val poolConfig: ContainerPoolConfig = loadConfigOrThrow[ContainerPoolConfig](ConfigKeys.containerPool)
+
+  val invokerName = InvokerInstanceId(0, None, None, poolConfig.userMemory)
+
+  /** 1. Publish a message to the loadbalancer */
+  override def publish(action: ExecutableWhiskActionMetaData, msg: ActivationMessage)(
+    implicit transid: TransactionId): Future[Future[Either[ActivationId, WhiskActivation]]] = {
+
+    /** 2. Update local state with the activation to be executed scheduled. */
+    val activationResult = setupActivation(msg, action, invokerName)
+    sendActivationToInvoker(messageProducer, msg, invokerName).map(_ => activationResult)
+  }
+
+  /** Creates an invoker for executing user actions. There is only one invoker in the lean model. */
+  private def makeALocalThreadedInvoker() {
+    implicit val ec = ExecutionContextFactory.makeCachedThreadPoolExecutionContext()
+    val actorSystema: ActorSystem =
+      ActorSystem(name = "invoker-actor-system", defaultExecutionContext = Some(ec))
+    new InvokerReactive(config, invokerName, messageProducer)(actorSystema, implicitly)
+  }
+
+  makeALocalThreadedInvoker()
+
+  override protected val invokerPool: ActorRef = actorSystem.actorOf(Props.empty)
+
+  override protected def releaseInvoker(invoker: InvokerInstanceId, entry: ActivationEntry) = {
+    // Currently do nothing
+  }
+
+  override protected def emitHistogramMetric() = {
+    // Currently do nothing
+  }
+}
+
+object LeanBalancer extends LoadBalancerProvider {
+
+  override def instance(whiskConfig: WhiskConfig, instance: ControllerInstanceId)(
+    implicit actorSystem: ActorSystem,
+    logging: Logging,
+    materializer: ActorMaterializer): LoadBalancer = {
+
+    new LeanBalancer(whiskConfig, createFeedFactory(whiskConfig, instance), instance)
+  }
+
+  def requiredProperties =
```

Comment:
i think you can remove service port form this - it can't be different from the controller's port and not sure it's used anyway is it?